pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1550 @@
+                                            debug("promise outdated", state.path);
+                                        }
+                                    }
+                                );
+                    this.#set_validation_state_promise(state, val, prom);
+                }
+            );
+        }
+    }
+
+    /* Since the DialogValidationState for a path is so important, it
+       is also important to keep them firmly associated with each
+       other when the path of a value changes.
+
+       A path might change when there are arrays involved, and
+       elements get new indices without actually changing identity.
+     */
+
+    _rename_validation_state(path: string, from: number, to: number) {
+        const from_path = path + "." + String(from);
+        const to_path = path + "." + String(to);
+        if (from_path in this.#validation_state) {
+            debug("rename", from_path, to_path);
+            this.#validation_state[to_path] = this.#validation_state[from_path];
+            this.#validation_state[to_path].path = to_path;
+            delete this.#validation_state[from_path];
+        }
+        for (const k in this.#validation_state) {
+            if (k.indexOf(from_path + ".") == 0) {
+                const to = to_path + k.substring(from_path.length);
+                debug("rename", k, to);
+                this.#validation_state[to] = this.#validation_state[k];
+                this.#validation_state[to].path = to;
+                delete this.#validation_state[k];
+            }
+        }
+    }
+
+    /* The first thing run_action does is to trigger a new validation
+       round and then wait for all the asynchronous results to have
+       come in.
+
+       If there are any DialogValidationState objects that are waiting
+       for a timeout, we want to cancel those and start over, so that
+       their validation starts immediately. (Also, it would be hairy
+       to wait for those timeouts to be over from here.)
+     */
+
+    async validate(): Promise<boolean> {
+        this.#cancel_all_validation_timeouts();
+        this.#trigger_validation();
+        await this.#wait_for_validation_promises();
+        return !this.#validation_failed;
+    }
+
+    #cancel_all_validation_timeouts() {
+        for (const p in this.#validation_state) {
+            const state = this.#validation_state[p];
+            if (state.timeout_id) {
+                debug("timeout bulk cancel", p);
+                window.clearTimeout(state.timeout_id);
+                delete this.#validation_state[p];
+            }
+        }
+    }
+
+    async #wait_for_validation_promises(): Promise<void> {
+        for (const path in this.#validation_state) {
+            const state = this.#validation_state[path];
+            if (state.promise) {
+                debug("waiting for promise", path);
+                await state.promise;
+                debug("waiting for promise done", path);
```

Comment:
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-0b143286-20251218-161050-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)